llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -3305,6 +3329,13 @@ RISCVTargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *AI) const {
   if (AI->isFloatingPointOperation())
     return AtomicExpansionKind::CmpXChg;
 
+  // The atomic RMW instructions are only available in capmode, so for the
+  // hybrid ABI we have expand these operations to use LR/SC (which are
+  // available even when not in capmode). However, the TLI hooks for
+  // AtomicExpansionKind::LLSC are not implemented so we use CmpXChg.
```

Comment:
Yes, because those are broken at -O0 due to stack spills being inserted in between, and are the whole reason we use late-expanded pseudos?

Reply:
And we don't need CmpXChg we can do the same thing as the existing code?

Reply:
ARM seems to always use it for RMW instructions and only disable it for cmpxchg expansion:

```
// For the real atomic operations, we have ldxr/stxr up to 128 bits,
TargetLowering::AtomicExpansionKind
AArch64TargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *AI) const {
  if (AI->isFloatingPointOperation())
    return AtomicExpansionKind::CmpXChg;

  unsigned Size = AI->getType()->getPrimitiveSizeInBits();
  if (Size > 128) return AtomicExpansionKind::None;
  // Nand not supported in LSE.
  if (AI->getOperation() == AtomicRMWInst::Nand) return AtomicExpansionKind::LLSC;
  // Leave 128 bits to LLSC.
  return (Subtarget->hasLSE() && Size < 128) ? AtomicExpansionKind::None : AtomicExpansionKind::LLSC;
}

TargetLowering::AtomicExpansionKind
AArch64TargetLowering::shouldExpandAtomicCmpXchgInIR(
    AtomicCmpXchgInst *AI) const {
  // If subtarget has LSE, leave cmpxchg intact for codegen.
  if (Subtarget->hasLSE())
    return AtomicExpansionKind::None;
  // At -O0, fast-regalloc cannot cope with the live vregs necessary to
  // implement cmpxchg without spilling. If the address being exchanged is also
  // on the stack and close enough to the spill slot, this can lead to a
  // situation where the monitor always gets cleared and the atomic operation
  // can never succeed. So at -O0 we need a late-expanded pseudo-inst instead.
  if (getTargetMachine().getOptLevel() == CodeGenOpt::None)
    return AtomicExpansionKind::None;
  return AtomicExpansionKind::LLSC;
}
```

Reply:
This is the RISC-V backend, not the AArch64 backend, so we should do it the same way as the RISC-V backend does for normal atomics.

Reply:
Yes and that's what I am doing, it's just a comment for potential future improvements.
Maybe in the future we (i.e. not me, since I don't care enough) would like to improve the generated code by not doing a cmpxchg and having a simpler LR/SC loop instead. The capability ops shouldn't break the forward progress guarantee so we could emit a simpler loop.